control_panel_api/permissions.py

```diff
@@ -93,24 +95,35 @@ def has_object_permission(self, request, view, obj):
 
 
 class UserS3BucketPermissions(BasePermission):
-    """
-    - Superusers can do anything
-    - Normal users can list/retrieve UserS3Buckets
-
-    NOTE: Filters are applied before permissions
-    """
-
     def has_permission(self, request, view):
         if is_superuser(request.user):
             return True
 
         if request.user.is_anonymous():
             return False
 
-        return view.action in ('list', 'retrieve')
+        if view.action != 'create':
+            return True
 
+        serializer = UserS3BucketSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
```

Comment:
Validate the data is not a "permissions" responsibility so it shouldn't be done here (it's already done in the views)

Reply:
Without this the bad request wouldn't even get to the view to be checked, it would just error out with a KeyError or type mismatch depending on how bad the submitted data is.

This is sort of a chicken and egg situation.  Still need to check the data as this permission check relies on existing data when creating.

Reply:
I see what you mean, potentially the input data may not have an `s3bucket` field, etc...